gsrs-module-substances-core/src/main/java/ix/ginas/utils/validation/validators/TagsValidator.java

```diff
@@ -0,0 +1,159 @@
+
+import java.util.*;
+@Slf4j
+
+/*
+
+=== Tags Validator Documentation ===
+
+The idea of substance tags appears in two ways in the GSRS. 1) in the list of names. 2) in the list of tags.
+
+A tag is included in a substance name by putting a bracketed term at the end of the name.
+For example:
+  ASPIRIN [INN]
+When this name is added, the GSRS will extract "INN" and index this value for faceting.
+The facet category called "Source Tag" works with this indexed value.
+There is ALSO a substance->tags list of Keyword objects.
+In the GSRS Frontend, tag values can be added manually to this list.
+Thus, there are two sources of truth for substance tags.
+This validator may be used to keep the lists consistent.
+
+This should be configured in your gsrs-ci gsrs-ci/substances/src/main/resources/application.conf.
+
+  # Without this setting, the validator will not be run on submission of a substance.
+  gsrs.validators.substances += {
+                "validatorClass" = "ix.ginas.utils.validation.validators.TagsValidator",
+                "newObjClass" = "ix.ginas.models.v1.Substance"
+        }
+
+  # These may be added, though defaults are set in the code.
+  gsrs.substance.addtagwhenintagspresentinnames=false
+  gsrs.substance.removetagwhenintagsmissingfromnames=false
+
+*** Note also that there is an idea of the locator.  This is used in conjunction with references.
+Currently this is OFF in GSRS (controlled by a boolean value extractLocators). If that were on,
+a method in NamesValidator.java (addLocator) would add tags found in names to the tags list. We should
+probably remove this piggybacking in case locators is ever turned back on.
+*/
+
+
+public class TagsValidator extends AbstractValidatorPlugin<Substance> {
+
+    @Value("#{new Boolean('${gsrs.substance.addtagwhenintagspresentinnames:false}')}")
+    private Boolean addTagWhenInTagsPresentInNames;
+
+    @Value("#{new Boolean('${gsrs.substance.removetagwhenintagsmissingfromnames:false}')}")
+    private Boolean removeTagWhenInTagsMissingFromNames;
+
+    @Autowired
+    private SubstanceRepository substanceRepository;
+
+    // Can't figure out how to set property in test; so doing this.
+    public void setAddTagWhenInTagsPresentInNames(Boolean b) {
+        this.addTagWhenInTagsPresentInNames = b;
+    }
+
+    // Can't figure out how to set property in test; so doing this.
+    public void setRemoveTagWhenInTagsMissingFromNames(Boolean b) {
+        this.removeTagWhenInTagsMissingFromNames = b;
+    }
+
+    @Override
+    public void validate(Substance s, Substance objold, ValidatorCallback callback) {
+        log.info("Inside Tags Validator.validate");
+
+        boolean skipRest = false;
+        if (removeTagWhenInTagsMissingFromNames == null) { removeTagWhenInTagsMissingFromNames = false; }
+
+        List<String> tagTermsFromNames = s.extractDistinctTagTermsFromNames();
+        List<String> tagTerms = s.grabTagTerms();
+
+        if (tagTermsFromNames == null) {
```

Comment:
why do this? It's not an error for tags to be empty ... I think you can remove this. Especially if you have the method return empty set when empty.